docs/design/process_group_crd.md

```diff
@@ -0,0 +1,103 @@
+Additional to the fields above in the `FDBProcessGroup` status it will contain the public IP address of the `FDBProcessGroup`, this can be the Pod IP or the Service IP.
+The `FDBProcessGroup` CRD will get some of fields that are currently represented in the FDB cluster CRD e.g. similar to a `ReplicaSet` that mirrors some of th fields from a `Deployment`:
+
+- The final [ProcessSettings](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#processsettings) for the specific process.
+- `instanceIDPrefix` for the created resources.
+- `updatePodsByReplacement`.
+- Part of the [ServiceConfig](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#serviceconfig) once we support to override fields in the created services.
+- `storageServersPerPod`.
+- `logGroup`.
+- `sidecarVariables`.
+
+If one of the fields above are changed this will trigger a reconcilation in the `FDBProcessGroup` for the specific sub-controller.
+Additional to the fields above that are taken from the `FDB cluster spec` we will add the fields from [ProcessGroupStatus](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#processgroupstatus) to the Spec:
+
+- `processClass`
+- `remove`
+- `excluded`
+- `exclusionSkipped`
+
+`addresses` is already included in the status field and the `processGroupID` should be used as the name of the `FDBProcessGroup`.
+Currently we have the following `ProcessGroupConditions` as part of the `ProcessGroupStatus`:
+
+```go
+const (
+	// NotConnecting represents a process group that doesn't connect to the cluster.
+	NotConnecting ProcessGroupConditionType = "NotConnecting"
+	// IncorrectPodSpec represents a process group that has an incorrect Pod spec.
+	IncorrectPodSpec ProcessGroupConditionType = "IncorrectPodSpec"
+	// IncorrectConfigMap represents a process group that has an incorrect ConfigMap.
+	IncorrectConfigMap ProcessGroupConditionType = "IncorrectConfigMap"
+	// IncorrectCommandLine represents a process group that has an incorrect commandline configuration.
+	IncorrectCommandLine ProcessGroupConditionType = "IncorrectCommandLine"
+	// PodFailing represents a process group which Pod keeps failing.
+	PodFailing ProcessGroupConditionType = "PodFailing"
+	// MissingPod represents a process group that doesn't have a Pod assigned.
+	MissingPod ProcessGroupConditionType = "MissingPod"
+	// MissingPVC represents a process group that doesn't have a PVC assigned.
+	MissingPVC ProcessGroupConditionType = "MissingPVC"
+	// MissingService represents a process group that doesn't have a Service assigned.
+	MissingService ProcessGroupConditionType = "MissingService"
+	// MissingProcesses represents a process group that misses a process.
+	MissingProcesses ProcessGroupConditionType = "MissingProcesses"
+)
+```
+
+Some of them are currently unused and most of them can simply be replaced by the `FDBProcessGroup` status.
```

Comment:
TODO: We need to define./figure out who is responsible for fetching the state and compare desired <-> current state.

Reply:
Yeah, I'm nervous about the potential for conflicts if we have two different reconciliation loops that are operating on the same status. On the other hand, some of these things depend on the running state of the cluster which means they wouldn't necessarily get updated properly in a per-process-group reconciliation loop.

Reply:
Maybe we could keep the logic for evaluating conditions in the process group reconciler, but have the cluster reconciler be able to call out to the process group reconciler in the cluster reconciler's UpdateStatus action to make sure the process group status is also up-to-date.

Reply:
The actual idea is to prevent these conflicts (see some lines below) where each controller only reads/writes either the Status or the Spec. I mean in the end thinking about it most of the status evaluation can be done by the cluster controller (so we only need to read the status once). With the information like `locality_id` the cluster controller is able to detect if a ProcessGroup has a missing process.